### PR TITLE
Add a Python script to automate stats downloading.

### PR DESCRIPTION
**Pull request by prbprbprb**
Downloads and formats stats, and can optionally update index.md with them.

NB will give slightly different results than the current script as it takes Android 14 numbers into account.  These are not currently included in the table, but will accumulate into Androd 13 etc.

### CODE REVIEW

**Comment by prbprbprb** [comment]
Gentle ping.

File: usage_reporter.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+import requests
+import urllib.parse
+from datetime import datetime
+from typing import Optional
+
+# Usage: python usage_reporter.py [update]
+#
+# If no update argument is given, downloads stats for the previous calendar month and
+# displays them.  NB relies on the fact that if only a single month is requested, the data
```
Comment (ebelinski) [inline usage_reporter.py:28]: Maybe a dumb question but what is "NB" in this context?

**Review by ebelinski — COMMENTED** [review]
Hey Pete, thank you for putting this together, this script is fantastic! 

I'm not sure what to make of the copyright statement and Apache license at the top though. This repository as a whole is licensed under [MIT](https://github.com/ebelinski/apilevels/blob/main/LICENSE). If we merge this PR as-is, would it mean that the script is licensed under Apache, and the rest of the repository would be under MIT? Would others still be allowed to make changes to this script in the future? I'm not familiar with the intricacies of open source licensing, and I'd be ok with merging as-is, but I just want to make sure that that would work.

Again, thanks for building this useful script and sorry for the delay in my response, I'll be more prompt moving forward. 🙂

File: usage_reporter.py (hunk 1 above)
Comment (TWiStErRob) [reply usage_reporter.py:28]: I think it's https://en.wikipedia.org/wiki/Nota_bene

**Comment by prbprbprb** [comment]
> I'm not sure what to make of the copyright statement and Apache license at the top though. 

Oops, my bad, force of habit as most of what I work on is Apache... I'll modify it to match the rest of the project.

File: usage_reporter.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    infilename = 'index.md'
+    tmpfilename = 'index.tmp'
+    level_re = re.compile('.*Level (\\d+)')
+    data_re = re.compile('(.*progress-cell.html.* percentage=)([0-9.]+)(.*)')
+    state = 'init'
+    api_level = None
+    with open(infilename, mode='r', encoding="utf-8") as infile:
+        with open(tmpfilename, mode='w', encoding="utf-8") as outfile:
+            for line in infile:
+                match state:
```
[the remark this answers is not in the document]
Comment (prbprbprb) [reply usage_reporter.py:125]: Good point... I actually don't work in Python much myself but I figured it's easily available so a good common language.  I'll find out the idiomatic way to do this and fail with useful diagnostics.

File: usage_reporter.py (hunk 1 above)
Comment (prbprbprb) [reply usage_reporter.py:28]: Will reword!

**Comment by prbprbprb** [comment]
Kinda busy today ;)  

But I'll address the comments this week.

File: usage_reporter.py (hunk 1 above)
Comment (prbprbprb) [reply usage_reporter.py]: Reworded.

File: usage_reporter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prbprbprb) [reply usage_reporter.py]: License fixed.

Unfortunately `match` is a new keyword in Python 3.10 so a runtime assertion won't work. And of course it makes sense to work with the default Python3 on MacOS, so I reworked it into a class based state machine instead which doesn't need `match`.  Tested with:
```
Python 3.9.6 (default, Oct  1 2023, 17:38:11)
[Clang 15.0.0 (clang-1500.0.40.1)] on darwin
```
from a fairly recent Xcode command line tools.

**Comment by prbprbprb** [comment]
Review comments addressed, but I do plan to iterate on this a bit more and improve it, notably it's missing any tests and because I bashed it out one evening it's not really structured in a testable way...

**Review by ebelinski — APPROVED** [review]
This is fantastic! I'll go ahead and merge it now, but feel free to follow up with a separate PR on tests and whatever else makes sense to you @prbprbprb.

**Pull request merged**